File: app/MMarket/js/src/data/Timetable.js
// Copyright 09-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Market open and close hours
**/
export default class Timetable {
  /**
      @param {number} hopen
      @param {number} mopen
      @param {number} hclose
      @param {number} mclose
   */
  constructor (hopen, mopen, hclose, mclose) {
    this._hopen = hopen;
    this._mopen = mopen;
    this._hclose = hclose;
    this._mclose = mclose;
  }

  /**
      @return {number}
  **/
  get hopen () {
    return this._hopen;
  }

  /**
      @return {number}
  **/
  get mopen () {
    return this._mopen;
  }

  /**
      @return {number}
  **/
  get hclose () {
    return this._hclose;
  }

  /**
      @return {number}
  **/
  get mclose () {
    return this._mclose;
  }

  /**
      @return {!Array<?>}
  **/
  toJs () {
    return [
      this._hopen,
      this._mopen,
      this._hclose,
      this._mclose
    ];
  }

  /**
      @param {!Array<?>} serial
      @return {!Timetable}
  **/
  static fromJs (serial) {
    return new Timetable(
      serial[0],
      serial[1],
      serial[2],
      serial[3]
    );
  }
}

File: app/MMarket_old/c/src/data/Timetable.c
// Copyright 16-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/Timetable.h"

/* .
# Market day time table.
-Timetable: serial
  hopen: int
  mopen: int
  hclose: int
  mclose: int
*/

/*--*/

struct Timetable_Timetable {
  int hopen;
  int mopen;
  int hclose;
  int mclose;
};

static Timetable *_timetable_new (
  int hopen,
  int mopen,
  int hclose,
  int mclose
) {
  Timetable *this = MALLOC(Timetable);
  this->hopen = hopen;
  this->mopen = mopen;
  this->hclose = hclose;
  this->mclose = mclose;
  return this;
}

int timetable_hopen (Timetable *this) {
  return this->hopen;
}

int timetable_mopen (Timetable *this) {
  return this->mopen;
}

int timetable_hclose (Timetable *this) {
  return this->hclose;
}

int timetable_mclose (Timetable *this) {
  return this->mclose;
}

Js *timetable_to_js (Timetable *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, js_wi((int)this->hopen));
  arr_push(js, js_wi((int)this->mopen));
  arr_push(js, js_wi((int)this->hclose));
  arr_push(js, js_wi((int)this->mclose));
  return js_wa(js);
}

Timetable *timetable_from_js (Js *js) {
  // Arr[Js]
  Arr *a = js_ra(js);
  Js **p = (Js **)arr_start(a);
  Timetable *this = MALLOC(Timetable);
  this->hopen = js_ri(*p++);
  this->mopen = js_ri(*p++);
  this->hclose = js_ri(*p++);
  this->mclose = js_ri(*p++);
  return this;
}

/*--*/

Timetable *timetable_new(void) {
  return _timetable_new(0, 0, 23, 55);
}
